src/core/components/hv-route/index.tsx

```diff
@@ -6,10 +6,152 @@
+  parser?: DomService.Parser;
+
+  navLogic: NavigatorService.Navigator;
+
+  constructor(props: Types.InnerRouteProps) {
+    super(props);
+
+    this.state = {
+      doc: null,
+      error: null,
+    };
+    this.navLogic = new NavigatorService.Navigator(this.props);
+  }
+
+  componentDidMount() {
+    this.parser = new DomService.Parser(
+      this.props.fetch,
+      this.props.onParseBefore || null,
+      this.props.onParseAfter || null,
+    );
+
+    this.load();
+  }
+
+  getUrl = (): string => {
+    return UrlService.getUrlFromHref(
+      this.props.url || this.props.entrypointUrl,
+      this.props.entrypointUrl,
+    );
+  };
+
+  /**
+   * Load the url and resolve the xml.
+   */
+  load = async (): Promise<void> => {
+    if (!this.parser) {
+      this.setState({
+        doc: null,
+        error: new NavigatorService.HvRouteError('No parser or context found'),
+      });
+      return;
+    }
+
+    try {
+      const url: string = this.getUrl();
+
+      const { doc } = await this.parser.loadDocument(url);
+      this.setState({
+        doc,
+        error: null,
+      });
+    } catch (err: unknown) {
+      this.setState({
+        doc: null,
+        error: err as Error,
+      });
+    }
+  };
+
+  /**
+   * View shown while loading
+   */
+  LoadingView = (): React.ReactElement => {
+    const LoadingScreen = this.props.loadingScreen || Loading;
+    return <LoadingScreen />;
+  };
+
+  /**
+   * View shown when there is an error
+   */
+  ErrorView = (props: { error: unknown }): React.ReactElement => {
+    const ErrorScreen = this.props.errorScreen || LoadError;
+    return (
+      <ErrorScreen
+        back={() => this.navLogic.back({})}
+        error={props.error}
+        onPressReload={() => this.load()}
+        onPressViewDetails={(uri: string | undefined) => {
+          this.navLogic.openModal({ url: uri });
+        }}
+      />
+    );
+  };
+
+  /**
+   * View shown when the document is loaded
+   */
+  ContentView = (): React.ReactElement => {
+    if (!this.props.url) {
+      throw new NavigatorService.HvRouteError('No url received');
+    }
+    if (!this.state.doc) {
+      throw new NavigatorService.HvRouteError('No document received');
+    }
+
+    const { ErrorView } = this;
+    try {
+      return (
+        <NavigatorService.Render
+          doc={this.state.doc}
+          element={this.props.element}
+          navLogic={this.navLogic}
+          routeProps={this.props}
+          url={this.getUrl()}
+        />
+      );
+    } catch (err) {
+      return <ErrorView error={err} />;
+    }
+  };
+
+  render() {
+    const { ErrorView, LoadingView, ContentView } = this;
+    if (this.state.error) {
+      return <ErrorView error={this.state.error} />;
+    }
+    if (this.state.doc) {
+      return <ContentView />;
+    }
+    return <LoadingView />;
+  }
+}
 
 /**
  * Functional component wrapper around HvRouteInner
```

Comment:
is there a particular reason why we need a inner component + a wrapper, instead of a single component?

Reply:
This was part of the rapid development where we decided against rewriting the whole thing. I can work on removing the class for this implementation

Reply:
The function provides access to multiple contexts. We had to wrap the class implementation inside the function to allow passing the multiple contexts' information in.